rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll

```diff
@@ -166,29 +166,18 @@ module SsaFlow {
   }
 }
 
-/**
- * Holds for expressions `e` that evaluate to the value of any last (in
- * evaluation order) subexpressions within it. E.g., expressions that propagate
- * a values from a subexpression.
- *
- * For instance, the predicate holds for if expressions as `if b { e1 } else {
- * e2 }` evalates to the value of one of the subexpressions `e1` or `e2`.
- */
-private predicate propagatesValue(Expr e) {
-  e instanceof IfExpr or
-  e instanceof LoopExpr or
-  e instanceof ReturnExpr or
-  e instanceof BreakExpr or
-  e.(BlockExpr).getStmtList().hasTailExpr() or
-  e instanceof MatchExpr
-}
-
 /**
  * Gets a node that may execute last in `n`, and which, when it executes last,
  * will be the value of `n`.
  */
-private ExprCfgNode getALastEvalNode(ExprCfgNode n) {
-  propagatesValue(n.getExpr()) and result.getASuccessor() = n
+private ExprCfgNode getALastEvalNode(ExprCfgNode e) {
+  e = any(IfExprCfgNode n | result = [n.getThen(), n.getElse()]) or
+  result = e.(LoopExprCfgNode).getLoopBody() or
+  result = e.(ReturnExprCfgNode).getExpr() or
+  result = e.(BreakExprCfgNode).getExpr() or
+  result = e.(BlockExprCfgNode).getTailExpr() or
+  result = e.(MatchExprCfgNode).getArmExpr(_) or
+  result.(BreakExprCfgNode).getTarget() = e
```

Comment:
Does this change/improve `getALastEvalNode` compared to before? The old implementation seems simpler to me.

Reply:
The idea is to move the responsibility for all CFG successor logic into `CfgNodes.qll`, so users of that library do not have to worry about the (low level) details of what the CFG looks like.

Reply:
So it's to not use `getASuccessor` in `DataFlowImpl.qll`?

I prefer the old implementation as it has fewer special cases, more closely matches the comment (except that the conjuncts are flipped), and separates the "expression that propagates values" property into a separate very simple predicate.

Reply:
It also makes it more explicit which child nodes we expect to propagate flow, for example the original version would incorrectly propagate flow from `b` to `if b { .. }` (when there is no `else` branch).

Reply:
That makes sense thanks. The old implementation is buggy in that regard.